### ISSUE TITLE
Need help debugging terrible performance

### ISSUE THREAD

**Issue by pretzelhammer** [SergioBenitez/Rocket#1631]
I implemented a RESTful HTTP JSON API server twice: [once using Diesel + Rocket](https://github.com/pretzelhammer/kanban/tree/main/diesel-rocket) and [a second time using sqlx + actix-web](https://github.com/pretzelhammer/kanban/tree/main/sqlx-actix-web). [I benchmarked both servers using the same tools and the same database](https://github.com/pretzelhammer/kanban/tree/main/benchmarks). These are the results I got:

- The **Diesel + Rocket** server had a max throughput of **198** requests per second.
- The **sqlx + actix-web** server had a max throughput of **2845** requests per second.

This difference is massive and the performance of the Diesel + Rocket server is terrible. [I asked for debugging help in the Diesel repo](https://github.com/diesel-rs/diesel/discussions/2755#discussioncomment-693482) and one of the maintainers told me Diesel absolutely should not be the cause of the terrible performance here because in their own benchmarks it's competitive with sqlx, so he suggested I make an issue on the Rocket repo which I'm doing now.

[The entire source code for the Diesel + Rocket server can be found here](https://github.com/pretzelhammer/kanban/tree/main/diesel-rocket). Preview of its Cargo.toml dependencies:

```toml
[dependencies]
dotenv = "0.15"
chrono = { version = "0.4", features = ["serde"] }
log = "0.4"
fern = "0.6"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
diesel = { version = "1.4", features = ["postgres", "chrono", "r2d2"] }
diesel-derive-enum = { version = "1.1", features = ["postgres"] }
rocket = "0.4"
rocket_contrib = "0.4"
```

[The entire source code for the sqlx + actix-web server can be found here](https://github.com/pretzelhammer/kanban/tree/main/sqlx-actix-web). Preview of its Cargo.toml dependencies:

```toml
[dependencies]
dotenv = "0.15"
chrono = { version = "0.4", features = ["serde"] }
log = "0.4"
fern = "0.6"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
sqlx = { version = "0.4", features = ["runtime-actix-rustls", "chrono", "postgres"] }
actix-web = "3.3"
futures = "0.3.14"
```

[The entire source code and instructions for how to run the benchmarks is here](https://github.com/pretzelhammer/kanban/tree/main/benchmarks). The main tools I used for benchmarking are vegeta and psutil. For vegeta I run a test for 60 secs and use up to 40 workers to generate load on the HTTP server. I have two test workloads, a read-only workload and a mixed (reads + writes) workload.

The read-only workload hits these endpoints:

```none
GET http://localhost:8000/api/boards
Authorization: Bearer LET_ME_IN

GET http://localhost:8000/api/boards/1/summary
Authorization: Bearer LET_ME_IN

GET http://localhost:8000/api/boards/1/cards
Authorization: Bearer LET_ME_IN
```

The mixed (reads + writes) workload hits these endpoints:

```none
GET http://localhost:8000/api/boards
Authorization: Bearer LET_ME_IN

GET http://localhost:8000/api/boards/1/summary
Authorization: Bearer LET_ME_IN

POST http://localhost:8000/api/boards
Authorization: Bearer LET_ME_IN
Content-Type: application/json
@post-board.json

DELETE http://localhost:8000/api/boards/1000000
Authorization: Bearer LET_ME_IN

GET http://localhost:8000/api/boards/1/cards
Authorization: Bearer LET_ME_IN

POST http://localhost:8000/api/cards
Authorization: Bearer LET_ME_IN
Content-Type: application/json
@post-card.json

PATCH http://localhost:8000/api/cards/1
Authorization: Bearer LET_ME_IN
Content-Type: application/json
@patch-card.json

DELETE http://localhost:8000/api/cards/1000000
Authorization: Bearer LET_ME_IN
```

Results from running the read-only workload on the Diesel + Rocket server:

```none
Requests      [total, rate, throughput]         13971, 231.22, 198.12
Duration      [total, attack, wait]             1m10s, 1m0s, 10.034s
Latencies     [min, mean, 50, 90, 95, 99, max]  6.178ms, 173.741ms, 27.678ms, 318.932ms, 326.678ms, 1.326s, 15.038s
Bytes In      [total, mean]                     2661222, 190.48
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           99.91%
Status Codes  [code:count]                      0:12  200:13959 
```

Results of running the mixed workload on the Diesel + Rocket server:

```none
Requests      [total, rate, throughput]         15455, 253.26, 188.32
Duration      [total, attack, wait]             1m6s, 1m1s, 4.981s
Latencies     [min, mean, 50, 90, 95, 99, max]  35.528µs, 164.5ms, 26.017ms, 313.756ms, 323.253ms, 1.428s, 14.772s
Bytes In      [total, mean]                     118362905, 7658.55
Bytes Out     [total, mean]                     205473, 13.29
Success       [ratio]                           80.43%
Status Codes  [code:count]                      0:3025  200:12430
```

Results of running the read-only workload on the sqlx + actix-web server:

```none
Requests      [total, rate, throughput]         170714, 2845.20, 2844.81
Duration      [total, attack, wait]             1m0s, 1m0s, 8.335ms
Latencies     [min, mean, 50, 90, 95, 99, max]  6.187ms, 14.026ms, 12.803ms, 17.904ms, 20.36ms, 27.41ms, 1.029s
Bytes In      [total, mean]                     32549337, 190.67
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:170714 
```

Results of running the mixed workload on the sqlx + actix-web server:

```none
Requests      [total, rate, throughput]         71522, 1191.92, 1189.90
Duration      [total, attack, wait]             1m0s, 1m0s, 101.888ms
Latencies     [min, mean, 50, 90, 95, 99, max]  6.706ms, 33.559ms, 21.452ms, 74.728ms, 96.839ms, 138.173ms, 1.169s
Bytes In      [total, mean]                     3515798166, 49156.88
Bytes Out     [total, mean]                     1385700, 19.37
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:71522 
```

As you can see from the above reports, the sqlx + actix-web server performance is pretty reasonable, but the Diesel + Rocket performance is absolutely abysmal. 

I compiled and ran both servers in release mode (Rust compiler 1.51 on Nightly), and I ran the benchmarks locally on my macOS machine, which has:
- CPU: 2.3 GHz 8-core Intel Core i9
- RAM: 32 GB 2667 MHz DDR4

Anyway, can someone [please take a look at the Diesel + Rocket server's source code](https://github.com/pretzelhammer/kanban/tree/main/diesel-rocket) and tell me if I'm doing something obviously wrong? I'm having a hard time believing these numbers because they're so bad, and there's no way Rocket can be _this slow_, right?

**Comment by SergioBenitez** [comment]
1) Make sure logging is disabled. Logging takes an incredible number of resources.
2) Significantly increase the Rocket worker count. Diesel blocks, so you're going to have at most k (the worker count) in-flight requests at once with Rocket + Diesel. SQLx is async, so you'll have as many in-flight requests as you want with SQLx + Actix.

**Comment by pretzelhammer** [comment]
1. I changed `log = "critical"` (the default) to `log = "off"` and that doesn't seem to have improved anything.
2. I changed `workers = 32` (the default for my machine) to `workers = 256` and that seems to have doubled the throughput performance, the Diesel + Rocket server can now process 350 - 400 requests per second, which is way more than before, but is still significantly less than the sqlx + actix-web server.
3. [Other than configuring the min idle connections and max total connections](https://github.com/pretzelhammer/kanban/blob/main/diesel-rocket/src/db.rs#L20-L26) I'm not sure what else I could change or do differently :/

> A more apples-to-apples comparison would be to compare `master` with sqlx.

Yeah, the current comparison isn't fair, and I wasn't expecting both servers to have equivalent performance. I thought that, at worst, the Diesel + Rocket server would perform 2x worse than the sqlx + actix-web server, but I was shocked when it performed 14x worse.

**Comment by SergioBenitez** [comment]
Even 256 will pale in comparison to starting thousands of tasks. Simply, different things are being measure. This particular workload - a fully blocking workload - is about worst-case for a fixed-thread model, as is 0.4. Again, you should see quite different results with `master`.

**Comment by pretzelhammer** [comment]
The reason I'm still confused is because I run the HTTP benchmark with 40 workers, and every worker sends requests serially (waits for a response before sending the next request, i.e. never sends more than 1 request at a time), so each HTTP server never has to handle more than 40 in-flight requests at once.

The connection pool is configured to aways hold a minimum of 50 idle connections, and a maximum of 100, so because there's at most only 40 requests that can happen in parallel I would expect zero contention for db connections and for the pool to always return a connection instantly without any blocking.

Also, with 256 worker threads, I would also expect the web framework to never "run out" in order to handle, at most, 40 requests in parallel.

Given the things above, I would expect a "sync" Rust web server to have performance in the same ballpark as an "async" Rust web server. I've read that that using an async runtime only pays off if the number of IO-bound tasks greatly outnumbers the number of CPU cores, and when asked how many IO-bound tasks is a lot people usually say "in the hundreds" or "in the thousands". But we're not talking about hundreds or thousands here, we're talking about just 40.

To make things really simple, I limited the number of workers to just 1. A single client, that sends requests, one at a time. The results:
- Diesel + Rocket: 155.31 req/sec over 1 minute
- sqlx + actix-web: 265.54 req/sec over 1 minute 

And here's the results of bumping it up to 5 workers:
- Diesel + Rocket: 227.83 req/sec over 1 minute
- sqlx + actix-web: 1136.75 req/sec over 1 minute 

Again, these numbers don't make any sense to me. I don't understand why there would be such a massive performance difference for such a tiny number of clients (each only sending 1 request at a time). But anyway, I've used up enough of your time, so feel free to ignore this comment.

**Comment by jebrosen** [comment]
One other thing you can try relatively easily is setting Rocket's keep_alive timeout to 0 to disable it, with either `keep_alive=0` in `Rocket.toml` or `ROCKET_KEEP_ALIVE=0` in the environment. Keep-alive in particular is broken in Rocket 0.4 (and, IIRC, much more noticeably in benchmarks than in real-world workloads), and unfortunately it seems like it won't be fixed because `hyper 0.10` is no longer actively maintained.
